File: the-method-of-programming/ch01-string/03-all-permutation/exercise-02.cpp
#include <cstring>
#include <iostream>

void calcAllCombinations(char *str) {
  const int base = strlen(str);
  const int max = base * 2 + 1;
  int n = 1;

  while (n <= max) {
    int i = 0;

    while (i < base) {
      if ((1 << i) & n) std::cout << str[i];
      ++i;
    }

    std::cout << std::endl;
    ++n;
  }
}

int main() {
  char str[] = "abc";
  calcAllCombination(str);

  return 0;
}

File: the-method-of-programming/ch01-string/03-all-permutation/exercise-02.js
function calcAllCombinations(str) {
  const base = str.length
  const max = base * 2 + 1
  let n = 1

  while (n <= max) {
    let i = 0
    let perm = ''
    while (i < base) {
      if ((1 << i) & n) perm += str[i]
      i++
    }
    console.log(perm)
    n++
  }
}

function main() {
  calcAllCombinations('abc')
}

main()
